workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,166 @@
+        private readonly Dictionary<Type, Dictionary<uint, IMemberAdapter>> adapterCache
+            = new Dictionary<Type, Dictionary<uint, IMemberAdapter>>();
+
+        // List of component IDs for which a reader/writer is required for each SpatialOSBehaviour
+        private readonly Dictionary<Type, List<uint>> componentReaderIdsForBehaviours = new Dictionary<Type, List<uint>>();
+        private readonly Dictionary<Type, List<uint>> componentWriterIdsForBehaviours = new Dictionary<Type, List<uint>>();
+
+        // Helper objects for querying members with the [Require] attribute
+        private readonly MemberReflectionUtil reflectionUtil = new MemberReflectionUtil(typeof(RequireAttribute));
+
+        private readonly HashSet<Type> badTypes = new HashSet<Type>();
+
+        private readonly ILogDispatcher logger;
+        private const string LoggerName = "SpatialOSBehaviourLibrary";
+
+        private const string BadRequiredMemberWarning
+            = "[Require] attribute found on member that is not Reader or Writer or incorrectly generated, ignoring!";
+
+        private const string MultipleReadersWritersRequiredError
+            = "MonoBehaviour found requesting more than one view (Reader or Writer) to the same component, " +
+            "this is invalid and will not be enabled!";
+
+        public SpatialOSBehaviourLibrary(ILogDispatcher logger)
+        {
+            this.logger = logger;
+        }
+
+        public void InjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
+            if (badTypes.Contains(spatialOSBehaviour.GetType()))
+            {
+                return;
+            }
+
+            foreach (var readerWriterComponentId in componentReaderIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                Inject(spatialOSBehaviour, readerWriterComponentId);
+            }
+
+            foreach (var readerWriterComponentId in componentWriterIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                Inject(spatialOSBehaviour, readerWriterComponentId);
+            }
+        }
+
+        public void DeInjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
+            if (badTypes.Contains(spatialOSBehaviour.GetType()))
+            {
+                return;
+            }
+
+            foreach (var readerWriterComponentId in componentReaderIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+
+            foreach (var readerWriterComponentId in componentWriterIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+        }
+
+        public IEnumerable<uint> GetRequiredReaderComponentIds(Type behaviourType)
```

Comment:
Is there a reason this returns an `IEnumerable` and not a `List` ?

Reply:
+1 - use concrete types unless there is a good reason not to

Reply:
Will change

Reply:
Could this be a hashset?

Reply:
Same comment as above. Are you iterating over it or sporadically checking if a value exists?

Reply:
Probably iterating so keeping it a list